### ISSUE TITLE
remove GCOVR macros

### ISSUE THREAD

**Issue by alyssawilk** [envoyproxy/envoy#19172]
they don't work and they're being misused.

tl;dr of thread below, they were added to mark code as not covered and exclude those lines from coverage builds.  They no longer exempt code from coverage.  They do crash if the line of code is unexpectedly hit in production. We've had at least one CVE due to an "uncovered" line being hit in production due to a code change and we'd like to avoid more.

The consensus was 
1) to switch all PANICing switch to have explicit cases statements, so that adding new fields will result in envoy not-building rather than building and PANICing in the odd corner case
2) and do a quick swap the GCOVR macros with 
  a) macros which make it clear the failure mode is PANIC in areas of code the GCOVR macro can't simply be removed(e.g. a PANIC in a function which returns a reference to a non-existant object)
  b) IS_ENVOY_BUG /ENVOY_BUG where we can skip crashing entirely (e.g. a void function)

once the PANICs are explicit, reviewers should push back about adding new crashes to Envoy.  Then https://github.com/envoyproxy/envoy/issues/19229 covers follow-up clean up of PANIC macros from both control and data plane.

**Comment by alyssawilk** [comment]
Ok have a first "easy" PR out, mostly replacing 'functions we don't expect to call' with ENVOY_BUG rather than production crash.

Going through GCOVR_EXCL uses, the by far most common one is 
switch (some enum)
case A:
// do work
case B:
// do more work
default:
PANIC() //crash the binary

which is great insofar as if you add a new field C, and don't correctly find every switch statement in the code base of that type, you introduce crashes.

I'm much more comfortable with
switch (some enum) {
case A:
// do work
case B:
// do more work
}
PANIC() //crash the only if there's corrupt data
so we compile-catch new additions, but that doesn't work for like 80% of the cases which are protobuf based.  Protos discourage exhaustive switch statements with "internal" sentinels: get compile errors of MATCH_PATTERN_NOT_SET and HttpConnectionManager_PathWithEscapedSlashesAction_HttpConnectionManager_PathWithEscapedSlashesAction_INT_MAX_SENTINEL_DO_NOT_USE_

For control plane validation having a default throw instead of default panic is clearly better.  Unfortunately it's not always clear which code is run from the control plane and which from the data plane.  cc @htuch for thoughts on replacing all proto enum switches with default throw as (IMO) it's better than panic and most (though not all) should be on the control plane.

**Comment by alyssawilk** [comment]
alternately we could have exhaustive case statements and a macro to panic if we see the sentinels?   I'm a big fan of exhaustive case statments...

**Comment by mattklein123** [comment]
> alternately we could have exhaustive case statements and a macro to panic if we see the sentinels? I'm a big fan of exhaustive case statments...

I would be a fan of this, I just don't know how we would enforce it other than code review?

**Comment by htuch** [comment]
The compiler would enforce the exhaustion property. I like the idea of exhaustive coverage inside the switch statement rather than relying on default, is there a way to make these macros compact and easy to use given the very verbose generated values? I'm sure there is a way, e.g. using protoc plugin to generate some aliases, but hopefully there is an easy way :)

**Comment by mattklein123** [comment]
> The compiler would enforce the exhaustion property.

Only if there is no default, which is something I often comment on in code reviews. It would be really nice if there was some way to tell the compiler that we don't allow default, but I don't think this exists.

**Comment by htuch** [comment]
> > The compiler would enforce the exhaustion property.
> 
> Only if there is no default, which is something I often comment on in code reviews. It would be really nice if there was some way to tell the compiler that we don't allow default, but I don't think this exists.

This would be a trivial clang-check plugin (if it doesn't already exist). Just need to scan the AST for forbidden default tokens inside switch.

**Comment by alyssawilk** [comment]
cool, I'lll work on setting up a macro and converting our existing code over.  Thanks to @RyanTheOptimist for the suggestion!
I think if we remove the GCOVR macro at least folks won't panic-by-default, and will likely mostly cargo cult code over.
